Slice syntax for calibration groups

kbwestfall opened the pull request:
The main purpose of this is to allow for slice syntax for calibration groups in the pypeit file.  The calibration group can now be 'all' or be some combination of listed numbers and slices.  Let's say you have 10 calibration groups.  Here are some examples of what you can do and what it implies:

 - 'all' = '0,1,2,3,4,5,6,7,8,9'
 - ':4' = '0,1,2,3'
 - '3:5,8:' = '3,4,8,9'
 - '3,1:5,6' = '1,2,3,4,6'

Note in that last one that it's smart enough to only keep unique elements in the list.

Along the way, I also:
- Streamlined some of the instantiation at the beginning of `PypeIt.__init__`.
    - Moves the call to `default_pypeit_par` into `config_specific_par`.
    - Adds a `finalize_usr_build` function to `PypeItMetaData` to consolidate the few opaque steps when finishing the meta data build.

Some tests fail, but I'm hoping they will pass when @jhennawi's current PR is merged.  Dev suite passes on Shane Kast.  I'll rerun the tests and the full dev suite once I've merged in his PR.

Note that this branch currently includes the differences between master and develop, so this also will update develop to be consistent with master.

profxj commented on pypeit/metadata.py:
```diff
@@ -1,10 +1,10 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import unicode_literals
+#from __future__ import print_function
```
These can be removed

kbwestfall replied:
Done.  And I just did all the rest of them, too...

profxj commented on pypeit/metadata.py:
```diff
@@ -223,202 +228,88 @@ def get_meta_data_model():
-            - `empty`: The columns are added but their values are all
-              originally set to -1.  **This is currently the only
-              option.**
+#    def _add_bkg_pairs(self, data, bkg_pairs):
```
Just checking:  `calib_id` and `bkg_id` don't show up for 
multi-slit, right?  Unless requested.

kbwestfall replied:
Right, I didn't change `set_pypeit_cols`.  These columns are always added to `fitstbl` just like before, but only output if `write_bkg_pairs` is true.

profxj commented on pypeit/metadata.py:
```diff
@@ -510,44 +401,49 @@ def merge(self, usrdata, match_type=True):
-        
-        msgs.error('Bad time unit')
-    '''
+        self.get_frame_types(user=frametype)
```
Thanks for cleaning this up!

kbwestfall replied:
NP

profxj commented on pypeit/metadata.py:
```diff
@@ -1328,22 +1286,29 @@ def set_pypeit_cols(self, write_bkg_pairs=False):
 
         """
-        # Set comb_id
+        if 'comb_id' not in self.keys():
```
Again, just checking that these tags don't show up in the .pypeit file
unless desired.

kbwestfall replied:
Yep.

profxj commented on pypeit/pypeit.py:
```diff
@@ -74,47 +74,51 @@ def __init__(self, pypeit_file, verbosity=2, overwrite=True, reuse_masters=False
-        # Write .calib file (For QA naming amongst other things)
+        #   - Interpret automated or user-provided data from the PypeIt
+        #   file
+        self.fitstbl.finalize_usr_build(frametype, setups[0])
```
Thanks for this.

kbwestfall replied:
NP.  Hopefully the doc string for `finalize_usr_build` is clear as to what it's doing.

kbwestfall commented:
@profxj Thanks for the quick comments!  I moved the main parsing code to `core/parse.py` and added a test to `test_parse.py`.  I'll wait to merge until I can merge in @jhennawi's PR.

kbwestfall commented:
Update: I've merged in both the current `master` and Joe's upcoming `cmartin_fixes` PR merge.  Tests pass.  I'll run the Keck dev-suite tonight.

profxj commented:
Ready to merge this?

kbwestfall commented:
Merging now.  Will update with master after coming PRs merge.
